File: android/app/src/main/java/app/notomorrow/feature/settings/SettingsAppEditors.kt
package app.notomorrow.feature.settings

import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.annotation.StringRes
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.unit.dp
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.compose.LifecycleEventEffect
import app.notomorrow.R
import app.notomorrow.designsystem.NT
import app.notomorrow.designsystem.NtText
import app.notomorrow.designsystem.PrimaryButton
import app.notomorrow.model.AIProvider
import app.notomorrow.push.rememberPushPermissionState
import app.notomorrow.service.HealthService
import app.notomorrow.util.S

// The "App" editors — `Features/Settings/SettingsAppEditors.swift`.

/** System / English / Polski. Applying the override recreates the Activity, so no relaunch hint. */
@Composable
fun LanguageEditor(
    state: SettingsUiState,
    model: SettingsViewModel,
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val current = SettingsFormat.language(state.languageOverride)
    StEditorScaffold(stringResource(S.settings_language), onBack, modifier) {
        StGroup(footnote = stringResource(R.string.settings_language_applies)) {
            LanguageOption.entries.forEach { option ->
                row {
                    StCheckRow(
                        title = stringResource(
                            when (option) {
                                LanguageOption.System -> S.settings_system
                                LanguageOption.English -> S.settings_english
                                LanguageOption.Polish -> S.settings_polish
                            }
                        ),
                        selected = current == option,
                        onClick = { model.setLanguage(option.code) },
                    )
                }
            }
        }
    }
}

/**
 * The two reminder toggles plus the real permission states: notifications, and — Android only —
 * the exact-alarm grant the rest countdown depends on.
 */
@Composable
fun NotificationsEditor(
    state: SettingsUiState,
    model: SettingsViewModel,
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val context = LocalContext.current
    val notifications = rememberPushPermissionState()
    val exactAlarms = rememberExactAlarmState()

    val permissionValue = stringResource(
        when {
            notifications.areNotificationsEnabled -> S.settings_notifications_allowed
            !notifications.canRequest -> S.settings_notifications_denied
            else -> S.settings_notifications_notAsked
        }
    )

    StEditorScaffold(stringResource(S.settings_notifications), onBack, modifier) {
        StGroup(footnote = stringResource(S.settings_notifications_footnote)) {
            row {
                StToggleRow(
                    title = stringResource(S.onboarding_schedule_remindHourBefore),
                    checked = state.remindHourBefore,
                    onCheckedChange = {
                        model.setRemindHourBefore(it)
                        if (it) notifications.request()
                    },
                )
            }
            row {
                StToggleRow(
                    title = stringResource(S.onboarding_schedule_askIfSkipped),
                    detail = stringResource(S.onboarding_schedule_askIfSkipped_detail),
                    checked = state.askIfSkippedAt21,
                    onCheckedChange = {
                        model.setAskIfSkipped(it)
                        if (it) notifications.request()
                    },
                )
            }
        }

        StGroup(title = stringResource(S.settings_notifications_permission)) {
            row {
                StInfoRow(
                    label = stringResource(S.settings_notifications_permission),
                    value = permissionValue,
                    dot = if (notifications.areNotificationsEnabled) NT.Colors.good else null,
                )
            }
            row {
                // iOS only swaps to the settings deep link on `.denied`; every other state keeps
                // offering "Request permission" (`SettingsAppEditors.swift:45-51`).
                val isDenied = !notifications.areNotificationsEnabled && !notifications.canRequest
                if (isDenied) {
                    StActionRow(
                        label = stringResource(S.settings_notifications_openSettings),
                        onClick = { openNotificationSettings(context) },
                    )
                } else {
                    StActionRow(
                        label = stringResource(S.settings_notifications_request),
                        onClick = notifications::request,
                    )
                }
            }
            // The rest countdown needs an exact alarm; the grant is a settings page, not a dialog.
            if (exactAlarmsAreRequestable) {
                row {
                    StInfoRow(
                        label = stringResource(R.string.settings_alarms_exact),
                        value = stringResource(
                            if (exactAlarms) S.settings_notifications_allowed
                            else S.settings_notifications_denied
                        ),
                        dot = if (exactAlarms) NT.Colors.good else null,
                    )
                }
                if (!exactAlarms) {
                    row {
                        StActionRow(
                            label = stringResource(R.string.settings_alarms_request),
                            onClick = { openExactAlarmSettings(context) },
                        )
                    }
                }
            }
        }
    }
}

/** Health Connect: unavailable / needs a provider update / connectable. */
@Composable
fun HealthEditor(
    state: SettingsUiState,
    model: SettingsViewModel,
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val context = LocalContext.current
    val permissions = remember(model) { model.healthPermissions }
    val launcher = rememberLauncherForActivityResult(HealthService.permissionsContract()) {
        model.refreshHealth()
    }
    LaunchedEffect(Unit) { model.refreshHealth() }
    LifecycleEventEffect(Lifecycle.Event.ON_RESUME) { model.refreshHealth() }

    val health = state.health

    StEditorScaffold(stringResource(S.settings_health), onBack, modifier) {
        NtText(
            text = stringResource(S.settings_health_description),
            style = NT.Fonts.subheadline,
            color = NT.Colors.ink2,
        )

        StGroup {
            row {
                StInfoRow(
                    label = stringResource(S.settings_health_status),
                    value = healthValue(health),
                    dot = if (health.isAuthorized) NT.Colors.good else null,
                )
            }
        }

        when {
            health.needsProviderUpdate -> {
                PrimaryButton(title = stringResource(R.string.settings_health_updateProvider)) {
                    openHealthConnectUpdate(context)
                }
            }

            health.isAvailable -> {
                PrimaryButton(title = stringResource(S.settings_health_connect)) {
                    launcher.launch(permissions)
                }
            }

            else -> {
                NtText(
                    text = stringResource(S.health_error_unavailable),
                    modifier = Modifier.padding(horizontal = 16.dp),
                    style = NT.Fonts.footnote,
                    color = NT.Colors.ink2,
                )
            }
        }

        if (health.error != null) {
            NtText(
                text = health.error,
                modifier = Modifier.padding(horizontal = 16.dp),
                style = NT.Fonts.footnote,
                color = NT.Colors.bad,
            )
        }
    }
}

/** Standard (Gemini through the backend), or Claude / Gemini with the user's own key (secure store). */
@Composable
fun AIProviderEditor(
    state: SettingsUiState,
    model: SettingsViewModel,
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
) {
    var keyDraft by remember { mutableStateOf("") }
    var geminiKeyDraft by remember { mutableStateOf("") }

    StEditorScaffold(stringResource(S.settings_aiProvider), onBack, modifier) {
        StGroup {
            row {
                StCheckRow(
                    title = stringResource(S.settings_ai_standard),
                    detail = stringResource(S.settings_ai_standard_detail),
                    selected = state.aiProvider == AIProvider.Standard,
                    onClick = { model.setAiProvider(AIProvider.Standard) },
                )
            }
            row {
                StCheckRow(
                    title = stringResource(S.settings_ai_claude),
                    detail = stringResource(S.settings_ai_claude_detail),
                    selected = state.aiProvider == AIProvider.ClaudeBYOK,
                    onClick = { model.setAiProvider(AIProvider.ClaudeBYOK) },
                )
            }
            row {
                StCheckRow(
                    title = stringResource(S.settings_ai_gemini),
                    detail = stringResource(S.settings_ai_gemini_detail),
                    selected = state.aiProvider == AIProvider.GeminiBYOK,
                    onClick = { model.setAiProvider(AIProvider.GeminiBYOK) },
                )
            }
        }

        // `.animation(.easeOut(0.2), value: model.aiProvider)` — the key section fades in.
        AnimatedVisibility(
            visible = state.aiProvider == AIProvider.ClaudeBYOK,
            enter = fadeIn(tween(200, easing = NT.Ease.out)),
            exit = fadeOut(tween(200, easing = NT.Ease.out)),
        ) {
            AIKeySection(
                masked = state.maskedAnthropicKey,
                labelRes = S.settings_ai_keyLabel,
                placeholderRes = S.settings_ai_keyPlaceholder,
                footnoteRes = S.settings_ai_footnote,
                draft = keyDraft,
                onDraft = { keyDraft = it },
                onSave = {
                    model.saveAnthropicKey(keyDraft)
                    keyDraft = ""
                },
                onRemove = {
                    model.removeAnthropicKey()
                    keyDraft = ""
                },
            )
        }

        AnimatedVisibility(
            visible = state.aiProvider == AIProvider.GeminiBYOK,
            enter = fadeIn(tween(200, easing = NT.Ease.out)),
            exit = fadeOut(tween(200, easing = NT.Ease.out)),
        ) {
            AIKeySection(
                masked = state.maskedGeminiKey,
                labelRes = S.settings_ai_geminiKeyLabel,
                placeholderRes = S.settings_ai_geminiKeyPlaceholder,
                footnoteRes = S.settings_ai_geminiFootnote,
                draft = geminiKeyDraft,
                onDraft = { geminiKeyDraft = it },
                onSave = {
                    model.saveGeminiKey(geminiKeyDraft)
                    geminiKeyDraft = ""
                },
                onRemove = {
                    model.removeGeminiKey()
                    geminiKeyDraft = ""
                },
            )
        }

        StGroup(title = stringResource(S.settings_developer)) {
            row {
                StToggleRow(
                    title = stringResource(S.settings_demoData),
                    detail = stringResource(S.settings_demoData_detail),
                    checked = state.useDemoData,
                    onCheckedChange = model::setUseDemoData,
                )
            }
        }
    }
}

/** `AIProviderEditor.keySection` — the stored key with a Remove row, or the secure field and Save. */
@Composable
private fun AIKeySection(
    masked: String?,
    @StringRes labelRes: Int,
    @StringRes placeholderRes: Int,
    @StringRes footnoteRes: Int,
    draft: String,
    onDraft: (String) -> Unit,
    onSave: () -> Unit,
    onRemove: () -> Unit,
) {
    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        if (masked != null) {
            StGroup {
                row {
                    StInfoRow(
                        label = stringResource(labelRes),
                        value = masked,
                        dot = NT.Colors.good,
                    )
                }
                row {
                    StActionRow(
                        label = stringResource(S.settings_ai_removeKey),
                        onClick = onRemove,
                        color = NT.Colors.bad,
                    )
                }
            }
        } else {
            StLabeled(stringResource(labelRes)) {
                StTextField(
                    value = draft,
                    onValueChange = onDraft,
                    placeholder = stringResource(placeholderRes),
                    keyboardOptions = KeyboardOptions(
                        autoCorrectEnabled = false,
                        keyboardType = KeyboardType.Password,
                        imeAction = ImeAction.Done,
                    ),
                    keyboardActions = KeyboardActions(onDone = { onSave() }),
                    visualTransformation = PasswordVisualTransformation(),
                )
            }
            PrimaryButton(
                title = stringResource(S.settings_ai_saveKey),
                height = NT.Size.cardButton,
                enabled = draft.trim().isNotEmpty(),
                onClick = onSave,
            )
        }

        NtText(
            text = stringResource(footnoteRes),
            modifier = Modifier.padding(horizontal = 16.dp),
            style = NT.Fonts.footnote,
            color = NT.Colors.ink2,
        )
    }
}

File: NoTomorrow/Features/Settings/SettingsAppEditors.swift
import SwiftUI
import SwiftData

// MARK: - Language

struct LanguageEditor: View {
    @Bindable var model: SettingsModel
    @Environment(AppState.self) private var appState

    var body: some View {
        STEditorScreen(title: "settings.language") {
            STGroup {
                STCheckRow(title: "settings.system", isSelected: appState.languageOverride == nil) { select(nil) }
                STCheckRow(title: "settings.english", isSelected: appState.languageOverride == "en") { select("en") }
                STCheckRow(title: "settings.polish", isSelected: appState.languageOverride == "pl") { select("pl") }
            }
            .stFootnote("settings.language.footnote")
        }
    }

    private func select(_ code: String?) {
        model.setLanguage(code, appState: appState)
    }
}

// MARK: - Notifications

struct NotificationsEditor: View {
    @Bindable var schedule: GymSchedule
    @Bindable var model: SettingsModel
    @Environment(\.scenePhase) private var scenePhase

    var body: some View {
        STEditorScreen(title: "settings.notifications") {
            STGroup {
                STToggleRow(title: "onboarding.schedule.remindHourBefore", isOn: remindHourBefore)
                STToggleRow(title: "onboarding.schedule.askIfSkipped",
                            detail: "onboarding.schedule.askIfSkipped.detail",
                            isOn: askIfSkipped)
            }
            .stFootnote("settings.notifications.footnote")

            STGroup(title: "settings.notifications.permission") {
                STInfoRow(label: "settings.notifications.permission", value: model.notificationStatusValue,
                          dot: model.notificationStatus == .authorized ? NT.Colors.good : nil)
                if model.notificationStatus == .denied {
                    STActionRow(label: "settings.notifications.openSettings") { model.openSystemSettings() }
                } else {
                    STActionRow(label: "settings.notifications.request") {
                        Task { await model.requestNotificationPermission() }
                    }
                }
            }
        }
        .task { await model.refreshNotificationStatus() }
        .onChange(of: scenePhase) { _, phase in
            if phase == .active { Task { await model.refreshNotificationStatus() } }
        }
    }

    private var remindHourBefore: Binding<Bool> {
        Binding(get: { schedule.remindHourBefore },
                set: { schedule.remindHourBefore = $0; schedule.updatedAt = .now })
    }

    private var askIfSkipped: Binding<Bool> {
        Binding(get: { schedule.askIfSkippedAt21 },
                set: { schedule.askIfSkippedAt21 = $0; schedule.updatedAt = .now })
    }
}

// MARK: - Health

struct HealthEditor: View {
    @Bindable var model: SettingsModel

    var body: some View {
        STEditorScreen(title: "settings.health") {
            Text("settings.health.description")
                .font(NT.Fonts.subheadline)
                .foregroundStyle(NT.Colors.ink2)
                .fixedSize(horizontal: false, vertical: true)

            STGroup {
                STInfoRow(label: "settings.health.status", value: model.healthValue,
                          dot: model.health.isAuthorized ? NT.Colors.good : nil)
            }

            if model.health.isAvailable {
                PrimaryButton(title: "settings.health.connect") {
                    Task { await model.health.requestAuthorization() }
                }
            } else {
                Text("health.error.unavailable")
                    .font(NT.Fonts.footnote)
                    .foregroundStyle(NT.Colors.ink2)
                    .padding(.horizontal, 16)
            }

            if let error = model.health.lastError {
                Text(error).font(NT.Fonts.footnote).foregroundStyle(NT.Colors.bad).padding(.horizontal, 16)
            }
        }
    }
}

// MARK: - AI estimates

/// Standard (Gemini through our backend), or Claude / Gemini with the user's own API key (Keychain).
struct AIProviderEditor: View {
    @Bindable var model: SettingsModel
    @State private var keyDraft = ""
    @State private var geminiKeyDraft = ""
    @FocusState private var keyFocused: Bool
    @FocusState private var geminiKeyFocused: Bool

    private var provider: Binding<AIProvider> {
        Binding(get: { model.aiProvider }, set: { model.aiProvider = $0 })
    }

    var body: some View {
        STEditorScreen(title: "settings.aiProvider") {
            STGroup {
                STCheckRow(title: "settings.ai.standard", detail: "settings.ai.standard.detail",
                           isSelected: model.aiProvider == .standard) { provider.wrappedValue = .standard }
                STCheckRow(title: "settings.ai.claude", detail: "settings.ai.claude.detail",
                           isSelected: model.aiProvider == .claudeBYOK) { provider.wrappedValue = .claudeBYOK }
                STCheckRow(title: "settings.ai.gemini", detail: "settings.ai.gemini.detail",
                           isSelected: model.aiProvider == .geminiBYOK) { provider.wrappedValue = .geminiBYOK }
            }

            if model.aiProvider == .claudeBYOK {
                keySection
            }

            if model.aiProvider == .geminiBYOK {
                geminiKeySection
            }

            STGroup(title: "settings.developer") {
                STToggleRow(title: "settings.demoData", detail: "settings.demoData.detail", isOn: demoData)
            }
        }
        .animation(.easeOut(duration: 0.2), value: model.aiProvider)
    }

    private var demoData: Binding<Bool> {
        Binding(get: { model.useDemoData }, set: { model.useDemoData = $0 })
    }

    private var keySection: some View {
        VStack(alignment: .leading, spacing: 12) {
            if let masked = model.auth.maskedAnthropicKey {
                STGroup {
                    STInfoRow(label: "settings.ai.keyLabel", value: masked, dot: NT.Colors.good)
                    STActionRow(label: "settings.ai.removeKey", color: NT.Colors.bad) {
                        model.auth.removeAnthropicKey()
                        keyDraft = ""
                    }
                }
            } else {
                SecureField(String(localized: "settings.ai.keyPlaceholder"), text: $keyDraft)
                    .font(NT.Fonts.body)
                    .foregroundStyle(NT.Colors.ink)
                    .textContentType(.password)
                    .autocorrectionDisabled()
                    .textInputAutocapitalization(.never)
                    .submitLabel(.done)
                    .focused($keyFocused)
                    .onSubmit(saveKey)
                    .stField(isFocused: keyFocused)
                    .stLabeled("settings.ai.keyLabel")

                PrimaryButton(title: "settings.ai.saveKey", height: NT.Size.cardButton,
                              isEnabled: !keyDraft.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty) {
                    saveKey()
                }
            }

            Text("settings.ai.footnote")
                .font(NT.Fonts.footnote)
                .foregroundStyle(NT.Colors.ink2)
                .fixedSize(horizontal: false, vertical: true)
                .padding(.horizontal, 16)
        }
    }

    private var geminiKeySection: some View {
        VStack(alignment: .leading, spacing: 12) {
            if let masked = model.auth.maskedGeminiKey {
                STGroup {
                    STInfoRow(label: "settings.ai.geminiKeyLabel", value: masked, dot: NT.Colors.good)
                    STActionRow(label: "settings.ai.removeKey", color: NT.Colors.bad) {
                        model.auth.removeGeminiKey()
                        geminiKeyDraft = ""
                    }
                }
            } else {
                SecureField(String(localized: "settings.ai.geminiKeyPlaceholder"), text: $geminiKeyDraft)
                    .font(NT.Fonts.body)
                    .foregroundStyle(NT.Colors.ink)
                    .textContentType(.password)
                    .autocorrectionDisabled()
                    .textInputAutocapitalization(.never)
                    .submitLabel(.done)
                    .focused($geminiKeyFocused)
                    .onSubmit(saveGeminiKey)
                    .stField(isFocused: geminiKeyFocused)
                    .stLabeled("settings.ai.geminiKeyLabel")

                PrimaryButton(title: "settings.ai.saveKey", height: NT.Size.cardButton,
                              isEnabled: !geminiKeyDraft.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty) {
                    saveGeminiKey()
                }
            }

            Text("settings.ai.geminiFootnote")
                .font(NT.Fonts.footnote)
                .foregroundStyle(NT.Colors.ink2)
                .fixedSize(horizontal: false, vertical: true)
                .padding(.horizontal, 16)
        }
    }

    private func saveKey() {
        let trimmed = keyDraft.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        model.auth.anthropicKey = trimmed
        keyDraft = ""
        keyFocused = false
    }

    private func saveGeminiKey() {
        let trimmed = geminiKeyDraft.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        model.auth.geminiKey = trimmed
        geminiKeyDraft = ""
        geminiKeyFocused = false
    }
}
